aldryn_faq/views.py

```diff
@@ -2,165 +2,215 @@
+        declare your obj like this:
 
+            MyView(…, AllowPKsTooMixin, TranslatableSlugMixin, DetailView):
+        """
+        if self.pk_url_kwarg in self.kwargs:
+            return super(DetailView, self).get_object(queryset)
+
+        # OK, just let Parler have its way with it.
+        return super(AllowPKsTooMixin, self).get_object(queryset)
+
+
+# TODO: Move this to Aldryn Translation Tools
+class CanonicalUrlMixin(object):
+    """
+    Provides configurable control over how non-canonical URLs to views are
+    handled. A view can specify by setting 'non_canonical_url_response_type' to
+    one of 200, 301, 302 or 404. By default, handling will be to temporarily
+    redirect to the canonical URL.
+    """
+    non_canonical_url_response_type = 302
+
+    def get_non_canonical_url_response_type(self):
+        response_type = getattr(self, "non_canonical_url_response_type", None)
+        if response_type and response_type in [200, 301, 302, 404]:
+            return response_type
+        else:
+            return self.non_canonical_url_response_type
 
-class FaqCategoryMixin(FaqMixin):
+    def get(self, request, *args, **kwargs):
+        """
+        On GET, if the URL used is not the correct one, handle according to
+        preferences by either:
+            Allowing (200),
+            Temporarily redirecting (302),
+            Permanently redirecting (301) or
+            Failing (404).
+        """
+        if not hasattr(self, 'object'):
+            self.object = self.get_object()
+        url = self.object.get_absolute_url()
+        response_type = self.get_non_canonical_url_response_type()
+        if response_type == 200 or request.path == url:
+            return super(CanonicalUrlMixin, self).get(
+                request, *args, **kwargs)
+        if response_type == 302:
+            return HttpResponseRedirect(url)
+        elif response_type == 301:
+            return HttpResponsePermanentRedirect(url)
+        else:
+            raise Http404('This is not the canonical uri of this object.')
+
+
+class FaqCategoryMixin(AppConfigMixin):
+    """
+    Provides support for getting the category from the URL, even if the view
+    using it is a DetailView for a different model, or not a DetailView at all.
+    """
+    category_pk_url_kwarg = 'category_pk'
+    category_slug_url_kwarg = 'category_slug'
 
-    def dispatch(self, *args, **kwargs):
-        try:
-            return super(FaqCategoryMixin, self).dispatch(*args, **kwargs)
-        except OldCategoryFormatUsed as error:
-            return self.handle_old_url_exception(error)
+    def dispatch(self, request, *args, **kwargs):
+        self.current_language = get_language_from_request(
+            request, check_path=True)
+        return super(FaqCategoryMixin, self).dispatch(
+            request, *args, **kwargs)
 
-    def get_category_or_404(self):
+    def get_language(self):
```

Comment:
this should look at the request language, and makes more sense to do it in dispatch() like self.language = ''